packages/googlesheets/src/Adaptor.js

```diff
@@ -79,20 +112,113 @@ export function appendValues(params) {
-            console.log(err);
+            logError(err);
             reject(err);
           } else {
             console.log('Success! Here is the response from Google:');
-            console.log(response);
-            resolve(state);
+            console.log(response.data);
+            resolve(
+              callback({
+                ...composeNextState(state, response.data),
+                response,
+              })
+            );
           }
         }
       );
     });
   };
 }
 
+/**
+ * Batch Updates values in a Spreadsheet.
+ * @example
+ * batchUpdateValues({
+ *   spreadsheetId: '1O-a4_RgPF_p8W3I6b5M9wobA3-CBW8hLClZfUik5sos',
+ *   range: 'Sheet1!A1:E1',
+ *   values: [
+ *     ['From expression', '$15', '2', '3/15/2016'],
+ *     ['Really now!', '$100', '1', '3/20/2016'],
+ *   ],
+ * })
+ * @function
+ * @param {Object} params - Data object to add to the spreadsheet.
+ * @param {string} [params.spreadsheetId] The spreadsheet ID.
+ * @param {string} [params.range] The range of values to update.
+ * @param {string} [params.valueInputOption] (Optional) Value update options. Defaults to 'USER_ENTERED'
+ * @param {array} [params.values] A 2d array of values to update.
+ * @returns {Operation} spreadsheet information
+ */
+export function batchUpdateValues(params, callback = s => s) {
+  return async state => {
+    const [resolvedParams] = expandReferences(state, params);
+
+    const {
+      spreadsheetId,
+      range,
+      valueInputOption = 'USER_ENTERED',
+      values,
+    } = resolvedParams;
+
+    const resource = {
+      data: [
+        {
+          range,
+          values,
+        },
+      ],
+      valueInputOption,
+    };
+    try {
+      const response = await client.spreadsheets.values.batchUpdate({
+        spreadsheetId,
+        resource,
+      });
+      console.log('%d cells updated.', response.data.totalUpdatedCells);
+      return callback({ ...composeNextState(state, response.data), response });
+    } catch (err) {
+      logError(err);
+      throw err;
+    }
+  };
+}
+
+/**
+ * Gets cell values from a Spreadsheet.
+ * @public
+ * @example
+ * getValues('1O-a4_RgPF_p8W3I6b5M9wobA3-CBW8hLClZfUik5sos','Sheet1!A1:E1')
+ * @function
+ * @param {string} spreadsheetId The spreadsheet ID.
+ * @param {string} range The sheet range.
+ * @returns {Operation} spreadsheet information
+ */
+export function getValues(spreadsheetId, range, callback = s => s) {
+  return async state => {
+    const [resolvedSheetId, resolvedRange] = expandReferences(
+      state,
+      spreadsheetId,
+      range
+    );
+
+    try {
+      const response = await client.spreadsheets.values.get({
+        spreadsheetId: resolvedSheetId,
+        range: resolvedRange,
+      });
+      const numRows = response?.data?.values.length ?? 0;
+      console.log(`${numRows} rows retrieved.`);
+
+      const nextState = { ...composeNextState(state, response.data), response };
```

Comment:
I also wonder if large data arrays are double-buffered here in `state.data` and `state.response`.

This might be a difficult situation I've stumbled into :grimacing:

Reply:
Do you mean duplicates ? or what do you mean by double-buffered ? @josephjclark